### DISCUSSION TITLE
BULK RAW with TinyUSB

### DISCUSSION THREAD

**Discussion by Prigana** [hathach/tinyusb#1708] [Q&A] [answered]
Hello, 
I would like to set up a bulk RAW using TinyUSB. So I managed to design a custom USB descriptor with one interface and two endpoints (BULK IN and OUT ).

![image](https://user-images.githubusercontent.com/95745240/198969944-28abbecd-2994-47c8-8728-c6b05cfe38f2.png)

To achieve this, I was inspired by the example using the CDC class in usb_descriptors.c .

![image](https://user-images.githubusercontent.com/95745240/198970271-f65bb979-57c6-4e58-91a1-c5e1a3549488.png)

(I kept the notification EPs but I will delete it in the future)

Then, I wanted to modify tusb_config.h to go through a custom class to do BULK RAW but it seems that the possibility of creating custom classes has not been implemented.

![image](https://user-images.githubusercontent.com/95745240/198971005-130e70b7-efbd-4f05-a06e-b30a478298ba.png)

So my problem is that when I call the tud_task() routine in main.c, my program crashes (my leds don't blink anymore).

![image](https://user-images.githubusercontent.com/95745240/198972034-21bd9eb2-3acb-4a7f-9a8d-68697e0cf012.png)

Can anyone help me? Has anyone done this type of project before? Can I create a custom class? How is the transfer in BULK mode handled in tinyUSB ? (which functions?)

**Comment by manuelbl** [comment, marked as answer]
The best approach likely is to use the *vendor* class.

Use the [webusb_serial example](https://github.com/hathach/tinyusb/tree/master/examples/device/webusb_serial) as a start and strip it down:

- Remove everything related to CDC
- Remove everything related to WebUSB

**Reply by Prigana** [reply]
Thank you @manuelbl so I used the vendor class correctly. I managed to do bulk with a python script using the pyusb library under ubuntu. However, when I want to switch to windows with a c++ script using the libusb library, it no longer works. I manage to claim my interface but not to send or receive data because each time, i have a timeout.

**Reply by Prigana** [reply]
# **My python script:**

#!/usr/bin/python

import usb.core
import usb.util

#find our device with VID And PID
dev = usb.core.find(idVendor=0x0483, idProduct=0x5740 )

#was it found?
if dev is None:
    raise ValueError('Device not found')

#set the active configuration. With no arguments, the first
#configuration will be the active one
dev.set_configuration(1)

#get an endpoint instance
cfg = dev.get_active_configuration()

intf = cfg[(0,0)]

#print intf

ep = usb.util.find_descriptor(
    intf,
    # match the first IN endpoint
    custom_match = \
	lambda e: \
		usb.util.endpoint_direction(e.bEndpointAddress) == \
		usb.util.ENDPOINT_IN)
ep2 = usb.util.find_descriptor(
    intf,
    # match the first OUT endpoint
    custom_match = \
	lambda e: \
		usb.util.endpoint_direction(e.bEndpointAddress) == \
		usb.util.ENDPOINT_OUT)
	
assert ep is not None
assert ep2 is not None

print ('ep:',ep)
print ('ep2:',ep2)

chunkr = 2**16
chunkw = 2**16

print ("Starting SCANNER python bulk test\n\r")

#Send DATA to endpoint out
ep2.write("HELLO")

#Receive DATA from endpoint in
data = dev.read(0x83, chunkw, 10)
print("\r\n")
print(data)

#Send DATA to endpoint out
ep2.write("BYE")

#Receive DATA from endpoint in
data = dev.read(0x83, chunkw, 10)
print(data)

usb.util.dispose_resources(dev)

**Reply by Prigana** [reply]
# **My C script**

int usb_write(void)
{

	int n, ret;
	n = 0;
	//write transfer
	//probably unsafe to use n twice...
	ret = libusb_bulk_transfer(handle, USB_ENDPOINT_OUT, transferBuf,sizeof(transferBuf), &n, USB_TIMEOUT);
	//Error handling
	switch (ret) {
	case 0:
		printf("send %d bytes to device\n", n);
		return 0;
	case LIBUSB_ERROR_TIMEOUT:
		printf("ERROR in bulk write: %d Timeout\n", ret);
		break;
	case LIBUSB_ERROR_PIPE:
		printf("ERROR in bulk write: %d Pipe\n", ret);
		break;
	case LIBUSB_ERROR_OVERFLOW:
		printf("ERROR in bulk write: %d Overflow\n", ret);
		break;
	case LIBUSB_ERROR_NO_DEVICE:
		printf("ERROR in bulk write: %d No Device\n", ret);
		break;
	default:
		printf("ERROR in bulk write: %d\n", ret);
		break;

	}
	return -1;

}

int test_bulk(void) 
{

	libusb_init(&ctx);
	//Open Device with VendorID and ProductID
	handle = libusb_open_device_with_vid_pid(ctx,
		USB_VENDOR_ID, USB_PRODUCT_ID);
	if (!handle) {
		return BULK_NOT_FOUND;
	}

	int r = 1;

	//Claim Interface 0 from the device
	r = libusb_claim_interface(handle, 0);
	if (r < 0) {
		return BULK_INTERFACE_NOT_CLAIM;
	} 
	while (1) 
	{
		usb_write();
	}
	libusb_close(handle);
	libusb_exit(NULL);


	return BULK_SUCCEED;
}

**Reply by manuelbl** [reply]
Is 0x40 really the address of your OUT endpoint? That doesn't seem right.

**Reply by Prigana** [reply]
Sorry, I changed the address of the endpoint in the meantime (I correct on my previous message), the problem does not come from there.

**Reply by manuelbl** [reply]
You are not reading anything. So it probably writes until the device's buffer and Window's buffer are full. Then it stalls and times out. This is assuming you have some sort of loopback setup.

**Reply by Prigana** [reply]
No, because I define the data of my buffer before as a global variable

          
          #include "bulk.h"
          #include <stdio.h>
          #include <stdlib.h>x
          
          #include <signal.h>
          #include <libusb.h>
          
          static libusb_context* ctx = NULL;
          static libusb_device_handle* handle;
          
          static uint8_t receiveBuf[64];
          uint8_t transferBuf[10]="HELLO";

**Reply by Prigana** [reply]
My problem may be due to my main.c

This is an echo program:
![image](https://user-images.githubusercontent.com/95745240/199979265-6ecb188d-dc02-4f02-9ae9-befd0df6025c.png)

**Reply by manuelbl** [reply]
Given that you have only shown part of the code, it's very difficult to say why it doesn't work.

The device code above is certainly problematic. You should not read from the vendor interface unless you are certain that the write buffer is not full. I can't tell you what happens if you call `tud_vendor_write()` and the write buffer is full. It might discard the data, or it might go havoc.

**Reply by Prigana** [reply]
Sorry, I didn't want to lose you, so I'm going to show you a lot more. So I use TinyUSB on my processor in the SOC of my FPGA generated by LITEX. I want to communicate in bulk from a Qt interface developed in C++ to my board. At first I used the pyusb library to test my bulk for its simplicity. I use the 7 leds of my board to debug my  code.

(Ps: I didnt start to play with tud_vendor_control_xfer_cb) 

So my main.c:

      #include <stdlib.h>
      #include <stdio.h>
      
      #include <generated/csr.h>
      #include <generated/mem.h>
      #include <generated/git.h>
      #include <generated/luna_usb.h>
      
      #include <irq.h>
      #include <uart.h>
      
      #include <sleep.h>
      #include <flash.h>
      
      #include "tusb.h"
      //--------------------------------------------------------------------+
      // MACRO CONSTANT TYPEDEF PROTYPES
      //--------------------------------------------------------------------+
      typedef struct
      {
	      uint32_t address;
	      uint32_t length;
      } memory_offest;
      
      memory_offest const alt_offsets[] = {
	      {.address = 0x200000, .length = 0x600000}, /* Main Gateware */
	      {.address = 0x800000, .length = 0x400000}, /* Main Firmawre */
	      {.address = 0xC00000, .length = 0x400000}, /* Extra */
	      {.address = 0x000000, .length = 0x200000}  /* Bootloader */
      };
      
      enum
      {
	      BLINK_BULK_SLEEP,
	      BLINK_1,
	      BLINK_2,
	      BLINK_3,
	      BLINK_4,
	      BLINK_5,
	      BLINK_6,
      };
      
      static uint32_t blink_state = BLINK_BULK_SLEEP;
      
      static bool flash_command_seen = false;
      static bool bus_reset_received = false;
      
      /* Blink pattern
       * - 1000 ms : device should reboot
       * - 250 ms  : device not mounted
       * - 1000 ms : device mounted
       * - 2500 ms : device is suspended
       */
      
      
      // Current system tick timer.
      volatile uint32_t system_ticks = 0;
      #if CFG_TUSB_OS == OPT_OS_NONE
      uint32_t board_millis(void)
      {
	      return system_ticks;
      }
      #endif
      
      //--------------------------------------------------------------------+
      // Board porting API
      //--------------------------------------------------------------------+
      
      static void timer_init(void)
      {
	      // Set up our timer to generate an interrupt every millisecond.
	      timer0_reload_write(60 * 1000);
	      timer0_en_write(1);
	      timer0_ev_enable_write(1);
      
	      // Enable our timer's interrupt.
	      irq_setie(1);
	      irq_setmask((1 << TIMER0_INTERRUPT) | irq_getmask());
      }
      
      static void timer_isr(void)
      {
	      // Increment our total millisecond count.
	      ++system_ticks;
      }
      
      void app_isr(void)
      {
	      unsigned int irqs;
	      irqs = irq_pending() & irq_getmask();
      
      
	      // Dispatch USB events.
	      if (irqs & (1 << USB_DEVICE_CONTROLLER_INTERRUPT | 1 << USB_IN_EP_INTERRUPT | 1 << USB_OUT_EP_INTERRUPT | 1 << USB_SETUP_INTERRUPT))
	      {
		      tud_int_handler(0);
	      }
      
	      /* Monitor bus resets */
	      if(irqs & (1 << USB_DEVICE_CONTROLLER_INTERRUPT))
	      {
		      bus_reset_received = true;
	      }
      
	      // Dispatch timer events.
	      if (irqs & (1 << TIMER0_INTERRUPT))
	      {
		      timer0_ev_pending_write(timer0_ev_pending_read());
		      timer_isr();
	      }
      
	      // Dispatch UART events.
	      if (irqs & (1 << UART_INTERRUPT))
	      {
		      uart_isr();
	      }
      }
      
      void led_blinking_task(void);
      void bulk_task(void);
      uint8_t usb_rx_ready;
      
      int main(int i, char **c)
      {
	      /* Setup IRQ, needed for UART */
	      irq_setmask(0);
	      irq_setie(1);
	      uart_init();
	      /* Enable VccIo
	       * Specifically we need ch2 enabled for the USB ULPI.
	       * But the hardware requires that we configure them all
	       */
	      vccio_ch0_write(45000); // 1v8
	      vccio_ch1_write(45000); // 1v8
	      vccio_ch2_write(45000); // 1v8
	      msleep(10);
	      vccio_enable_write(1);
      
	      usb_device_controller_reset_write(1);
	      msleep(20);
	      usb_device_controller_reset_write(0);
	      msleep(20);
	      usb_device_controller_connect_write(1);
	      msleep(20);
	      usb_device_controller_connect_read(); 
        
	      timer_init();
	      
	      // init device stack on configured roothub port
	      tud_init(BOARD_TUD_RHPORT);
	      while (1)
	      {
		      tud_task(); // tinyusb device task
		      bulk_task();
    		      led_blinking_task();		
      
      
		      if(button_in_read()==0)//Button Reset
		      {
		      reset_out_write(1);
		      }
	      }
	      
	      return 0;
      }
      
      const uint16_t sine_pulse_fade[] = {0x000, 0x000, 0x000, 0x017, 0x04d, 0x09a, 0x0f4, 0x15a, 0x1c5, 0x235, 0x29f, 0x2fd, 0x354, 0x39a, 0x3cf, 0x3f1, 0x3ff, 0x3f1, 0x3cf, 0x39a, 0x354, 0x2fd, 0x29f, 0x235, 0x1c5, 0x15a, 0x0f4, 0x09a, 0x04d, 0x017, 0x000, 0x000};
      
      void led_blinking_task(void) //LED (I have 7 LEDs on my board, I use them for debugging)
      
	      static uint32_t start_ms = 0;
	      static int count;
      
	      // Blink every interval ms
	      if ((board_millis() - start_ms) < 40)
		      return; // not enough time
	      start_ms += 20;
	      count++;
      
	      volatile uint32_t *p = (uint32_t *)CSR_LEDS_OUT0_ADDR;
	      
	      switch (blink_state)
	      {
	      
	      case BLINK_BULK_SLEEP:
	      {
		      for (int i = 0; i < 7; i++)
		      {
			      p[i] = 0;
		      }
		      p[0] = ((uint32_t)sine_pulse_fade[(count) % 32] /2); // RED
		      p[3] = ((uint32_t)sine_pulse_fade[(count) % 32] /2); // RED
		      p[6] = ((uint32_t)sine_pulse_fade[(count) % 32] /2); // RED
	      }
	      break;
	      
	      case BLINK_0: //SWITCH ON LED 0
	      {
		      for (int i = 0; i < 7; i++)
		      {
			      p[i] = 0;
		      }
		      p[0] = ((uint32_t)sine_pulse_fade[(count) % 32] /2); // RED
      
	      }
	      break;
      
      
	      case BLINK_1: //SWITCH ON LED 1
	      {
		      for (int i = 0; i < 7; i++)
		      {
			      p[i] = 0;
		      }
		      p[1] = ((uint32_t)sine_pulse_fade[(count) % 32] /2); // RED
      
	      }
	      break;
      
	      case BLINK_2: //SWITCH ON LED 2
	      {
		      for (int i = 0; i < 7; i++)
		      {
			      p[i] = 0;
		      }
		      p[2] = ((uint32_t)sine_pulse_fade[(count) % 32] /2); // RED
      
	      }
	      break;
      
	      case BLINK_3: //SWITCH ON LED 3
	      {
		      for (int i = 0; i < 7; i++)
		      {
			      p[i] = 0;
		      }
		      p[3] = ((uint32_t)sine_pulse_fade[(count) % 32] /2); // RED
      
	      }
	      break;
      
	      case BLINK_4: //SWITCH ON LED 4
	      {
		      for (int i = 0; i < 7; i++)
		      {
			      p[i] = 0;
		      }
		      p[4] = ((uint32_t)sine_pulse_fade[(count) % 32] /2); // RED
      
	      }
	      break;
      
	      case BLINK_5: //SWITCH ON LED 5
	      {
		      for (int i = 0; i < 7; i++)
		      {
			      p[i] = 0;
		      }
		      p[5] = ((uint32_t)sine_pulse_fade[(count) % 32] /2); // RED
      
	      }
	      break;
      
      
	      case BLINK_6: //SWITCH ON LED 6
	      {
		      for (int i = 0;  i < 7; i++)
		      {
			      p[i] = 0;
		      }
		      p[6] = ((uint32_t)sine_pulse_fade[(count) % 32] /2); // RED
      
	      }
	      break;
      
	      default:
	      {
	      blink_state=BLINK_BULK_SLEEP;
	      }
	      break;
      }}
      
      //--------------------------------------------------------------------+
      // Device callbacks
      //--------------------------------------------------------------------+
      
      // Invoked when device is mounted
      void tud_mount_cb(void)
      {
	      blink_state=BLINK_1;
      }
      
      // Invoked when device is unmounted
      void tud_umount_cb(void)
      {
      
      
      }
      
      // Invoked when usb bus is suspended
      // remote_wakeup_en : if host allow us  to perform remote wakeup
      // Within 7ms, device must draw an average of current less than 2.5 mA from bus
      void tud_suspend_cb(bool remote_wakeup_en)
      {
	      blink_state=BLINK_5;
	      (void)remote_wakeup_en;
      }
      
      // Invoked when usb bus is resumed
      void tud_resume_cb(void)
      {
          blink_state=BLINK_6;
      }
      
      bool tud_vendor_control_xfer_cb(uint8_t rhport, uint8_t stage, tusb_control_request_t const * request)
      {
          blink_state=BLINK_2;
          return tud_control_status(rhport, request);
      }
      
      void bulk_task(void)
      {
	      if ( tud_vendor_available())
          {
            blink_state=BLINK_3;    
            uint8_t buf[64];
            uint32_t count = tud_vendor_read(buf, sizeof(buf));
       
            tud_vendor_write(buf,count);
            tud_vendor_flush();
          }
      }
      
      void tud_vendor_rx_cb(uint8_t itf)
      {
          blink_state=BLINK_4;
      }

my usb_descriptors.c:

          #include "tusb.h"
          
          
          /* A combination of interfaces must have a unique product id, since PC will save device driver after the first plug.
           * Same VID/PID with different interface e.g MSC (first), then CDC (later) will possibly cause system error on PC.
           *
           * Auto ProductID layout's Bitmap:
           *   [MSB]         HID | MSC | CDC          [LSB]
           */
          #define USB_PID           0x0209
          
          #define USB_VID   0x0483
          #define USB_BCD   0x0200
          
          //--------------------------------------------------------------------+
          // Device Descriptors
          //--------------------------------------------------------------------+
          tusb_desc_device_t const desc_device =
          {
            .bLength            = sizeof(tusb_desc_device_t),
            .bDescriptorType    = TUSB_DESC_DEVICE,
            .bcdUSB             = USB_BCD,
          
            // Use Interface Association Descriptor (IAD) for CDC
            // As required by USB Specs IAD's subclass must be common class (2) and protocol must be IAD (1)
            .bDeviceClass       = 0X00,
            .bDeviceSubClass    = 0X00,
            .bDeviceProtocol    = 0X00,
          
            .bMaxPacketSize0    = CFG_TUD_ENDPOINT0_SIZE,
          
            .idVendor           = USB_VID,
            .idProduct          = USB_PID,
            .bcdDevice          = 0x0100,
          
            .iManufacturer      = 0x01,
            .iProduct           = 0x02,
            .iSerialNumber      = 0x03,
          
            .bNumConfigurations = 0x01
          };
          
          // Invoked when received GET DEVICE DESCRIPTOR
          // Application return pointer to descriptor
          uint8_t const * tud_descriptor_device_cb(void)
          {
            return (uint8_t const *) &desc_device;
          }
          
          //--------------------------------------------------------------------+
          // Configuration Descriptor
          //--------------------------------------------------------------------+
          
          enum
          {
            ITF_NUM_VENDOR = 0,
            ITF_NUM_TOTAL
          };
          
          
          #define EPNUM_VENDOR_IN  3
          #define EPNUM_VENDOR_OUT 3
          
          
          #define CONFIG_TOTAL_LEN    (TUD_CONFIG_DESC_LEN + TUD_VENDOR_DESC_LEN)
          
          // full speed configuration
          uint8_t const desc_fs_configuration[] =
          {
            // Config number, interface count, string index, total length, attribute, power in mA
            TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, 0x00, 100),
          
            // Interface number, string index, EP Out & IN address, EP size
            TUD_VENDOR_DESCRIPTOR(ITF_NUM_VENDOR, 5, EPNUM_VENDOR_OUT, 0x80 | EPNUM_VENDOR_IN, 64)
          
          };
          
          #if TUD_OPT_HIGH_SPEED
          // Per USB specs: high speed capable device must report device_qualifier and other_speed_configuration
          
          // high speed configuration
          uint8_t const desc_hs_configuration[] =
          {
            // Config number, interface count, string index, total length, attribute, power in mA
            TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, 0x00, 100),
          
            // Interface number, string index, EP Out & IN address, EP size
            TUD_VENDOR_DESCRIPTOR(ITF_NUM_VENDOR, 5, EPNUM_VENDOR_OUT, 0x80 | EPNUM_VENDOR_IN,512)
            
          };
          
          // other speed configuration
          uint8_t desc_other_speed_config[CONFIG_TOTAL_LEN];
          
          // device qualifier is mostly similar to device descriptor since we don't change configuration based on speed
          tusb_desc_device_qualifier_t const desc_device_qualifier =
          {
            .bLength            = sizeof(tusb_desc_device_qualifier_t),
            .bDescriptorType    = TUSB_DESC_DEVICE_QUALIFIER,
            .bcdUSB             = USB_BCD,
          
            .bDeviceClass       = 0X00,
            .bDeviceSubClass    = 0X00,
            .bDeviceProtocol    = 0X00,
          
            .bMaxPacketSize0    = CFG_TUD_ENDPOINT0_SIZE,
            .bNumConfigurations = 0x01,
            .bReserved          = 0x00
          };
          
          // Invoked when received GET DEVICE QUALIFIER DESCRIPTOR request
          // Application return pointer to descriptor, whose contents must exist long enough for transfer to complete.
          // device_qualifier descriptor describes information about a high-speed capable device that would
          // change if the device were operating at the other speed. If not highspeed capable stall this request.
          uint8_t const* tud_descriptor_device_qualifier_cb(void)
          {
            return (uint8_t const*) &desc_device_qualifier;
          }
          
          // Invoked when received GET OTHER SEED CONFIGURATION DESCRIPTOR request
          // Application return pointer to descriptor, whose contents must exist long enough for transfer to complete
          // Configuration descriptor in the other speed e.g if high speed then this is for full speed and vice versa
          uint8_t const* tud_descriptor_other_speed_configuration_cb(uint8_t index)
          {
            (void) index; // for multiple configurations
          
            // if link speed is high return fullspeed config, and vice versa
            // Note: the descriptor type is OHER_SPEED_CONFIG instead of CONFIG
            memcpy(desc_other_speed_config,
                   (tud_speed_get() == TUSB_SPEED_HIGH) ? desc_fs_configuration : desc_hs_configuration,
                   CONFIG_TOTAL_LEN);
          
            desc_other_speed_config[1] = TUSB_DESC_OTHER_SPEED_CONFIG;
          
            return desc_other_speed_config;
          }
          
          #endif // highspeed
          
          
          // Invoked when received GET CONFIGURATION DESCRIPTOR
          // Application return pointer to descriptor
          // Descriptor contents must exist long enough for transfer to complete
          uint8_t const * tud_descriptor_configuration_cb(uint8_t index)
          {
            (void) index; // for multiple configurations
          
          #if TUD_OPT_HIGH_SPEED
            // Although we are highspeed, host may be fullspeed.
            return (tud_speed_get() == TUSB_SPEED_HIGH) ?  desc_hs_configuration : desc_fs_configuration;
          #else
            return desc_fs_configuration;
          #endif
          }
          
          //--------------------------------------------------------------------+
          // String Descriptors
          //--------------------------------------------------------------------+
          
          // array of pointer to string descriptors
          char const* string_desc_arr [] =
          {
            (const char[]) { 0x09, 0x04 }, // 0: is supported language is English (0x0409)
            "LOOTUS",                     // 1: Manufacturer
            "SCANNER Device",              // 2: Product
            "123456789012",                // 3: Serials, should use chip ID
            "TinyUSB CDC",                 // 4: CDC Interface
          };
          
          static uint16_t _desc_str[32];
          
          // Invoked when received GET STRING DESCRIPTOR request
          // Application return pointer to descriptor, whose contents must exist long enough for transfer to complete
          uint16_t const* tud_descriptor_string_cb(uint8_t index, uint16_t langid)
          {
            (void) langid;
          
            uint8_t chr_count;
          
            if ( index == 0)
            {
              memcpy(&_desc_str[1], string_desc_arr[0], 2);
              chr_count = 1;
            }else
            {
              // Note: the 0xEE index string is a Microsoft OS 1.0 Descriptors.
              // https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/microsoft-defined-usb-descriptors
          
              if ( !(index < sizeof(string_desc_arr)/sizeof(string_desc_arr[0])) ) return NULL;
          
              const char* str = string_desc_arr[index];
          
              // Cap at max char
              chr_count = (uint8_t) strlen(str);
              if ( chr_count > 31 ) chr_count = 31;
          
              // Convert ASCII string into UTF-16
              for(uint8_t i=0; i<chr_count; i++)
              {
                _desc_str[1+i] = str[i];
              }
            }
          
            // first byte is length (including header), second byte is string type
            _desc_str[0] = (uint16_t) ((TUSB_DESC_STRING << 8 ) | (2*chr_count + 2));
          
            return _desc_str;
          }

As I already said previously my python script works (see capture below). However, if I restart my script a second time, my card crashes and it no longer works (The led stops flashing), I think I'm missing code allowing me to release my interface for the next run in my main.c.

![image](https://user-images.githubusercontent.com/95745240/200005057-963a928d-8706-454d-8a0c-517698a7f3a0.png)

My message transmitted in my script does echo twice. I can therefore do several echoes in a row but not several executions of my script in a row.

My tusb_config.h :

      /*
       * The MIT License (MIT)
       *
       * Copyright (c) 2019 Ha Thach (tinyusb.org)
       *
       * Permission is hereby granted, free of charge, to any person obtaining a copy
       * of this software and associated documentation files (the "Software"), to deal
       * in the Software without restriction, including without limitation the rights
       * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
       * copies of the Software, and to permit persons to whom the Software is
       * furnished to do so, subject to the following conditions:
       *
       * The above copyright notice and this permission notice shall be included in
       * all copies or substantial portions of the Software.
       *
       * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
       * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
       * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
       * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
       * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
       * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
       * THE SOFTWARE.
       *
       */
      
      #ifndef _TUSB_CONFIG_H_
      #define _TUSB_CONFIG_H_
      
      #ifdef __cplusplus
       extern "C" {
      #endif
      
      //--------------------------------------------------------------------+
      // Board Specific Configuration
      //--------------------------------------------------------------------+
      
      // RHPort number used for device can be defined by board.mk, default to port 0
      #ifndef BOARD_TUD_RHPORT
      #define BOARD_TUD_RHPORT      0
      #endif
      
      // RHPort max operational speed can defined by board.mk
      #ifndef BOARD_TUD_MAX_SPEED
      //#define BOARD_TUD_MAX_SPEED   OPT_MODE_FULL_SPEED
      #define BOARD_TUD_MAX_SPEED   OPT_MODE_HIGH_SPEED
      #endif
      
      //--------------------------------------------------------------------
      // Common Configuration
      //--------------------------------------------------------------------
      
      // defined by compiler flags for flexibility
      #ifndef CFG_TUSB_MCU
      #error CFG_TUSB_MCU must be defined
      #endif
      
      #ifndef CFG_TUSB_OS
      #define CFG_TUSB_OS           OPT_OS_NONE
      #endif
      
      #ifndef CFG_TUSB_DEBUG
      #define CFG_TUSB_DEBUG        0
      #endif
      
      // Enable Device stack
      #define CFG_TUD_ENABLED       1
      
      // Default is max speed that hardware controller could support with on-chip PHY
      #define CFG_TUD_MAX_SPEED     BOARD_TUD_MAX_SPEED
      
      /* USB DMA on some MCUs can only access a specific SRAM region with restriction on alignment.
       * Tinyusb use follows macros to declare transferring memory so that they can be put
       * into those specific section.
       * e.g
       * - CFG_TUSB_MEM SECTION : __attribute__ (( section(".usb_ram") ))
       * - CFG_TUSB_MEM_ALIGN   : __attribute__ ((aligned(4)))
       */
      #ifndef CFG_TUSB_MEM_SECTION
      #define CFG_TUSB_MEM_SECTION
      #endif
      
      #ifndef CFG_TUSB_MEM_ALIGN
      #define CFG_TUSB_MEM_ALIGN    __attribute__ ((aligned(4)))
      #endif
      
      //--------------------------------------------------------------------
      // DEVICE CONFIGURATION
      //--------------------------------------------------------------------
      
      #ifndef CFG_TUD_ENDPOINT0_SIZE
      #define CFG_TUD_ENDPOINT0_SIZE    64
      #endif
      
      //------------- CLASS -------------//
      #define CFG_TUD_CDC              0
      #define CFG_TUD_MSC              0
      #define CFG_TUD_HID              0
      #define CFG_TUD_MIDI             0
      #define CFG_TUD_VENDOR           1
      
      // Vendor FIFO size of TX and RX
      // If not configured vendor endpoints will not be buffered
      #define CFG_TUD_VENDOR_RX_BUFSIZE (TUD_OPT_HIGH_SPEED ? 512 : 64)
      #define CFG_TUD_VENDOR_TX_BUFSIZE (TUD_OPT_HIGH_SPEED ? 512 : 64)
      
      
      #ifdef __cplusplus
       }
      #endif
      
      #endif /* _TUSB_CONFIG_H_ */

**Comment by manuelbl** [comment]
So that's the device code. I was more interested in the host code.

You are saying it works with the Python script. The Python script is very trivial: send a few bytes, receive them again, and repeat once more. This will work even if the flow control on both the host and device side is broken.

The Windows code however is very different. It has an infinite loop sending data. It will quickly fill up all buffers and test your flow control, which is likely broken on the device side (as noted before). And it's still unclear to me if and how data is received on the host side.

Also note that your are likely sending garbage: `transferBuf ` is 10 bytes long and 10 bytes are sent. Yet `"HELLO"` only has 6 bytes including the terminating null.

Also note that USB bulk transfer is a stream (and not a message) protocol. So Windows will merge all submitted data, and the message boundaries are lost.

**Reply by Prigana** [reply]
Thank you for your feedback, it helps me a lot, I'm still a student, it explains the many blunders I can make.

Here is my C code, so I will modify it to avoid getting out of this infinite loop. This was not a problem for me as I used debug points after my libusb_bulk_transfer function calls to see error codes.

So I'm calling the test_bulk function in my Qt for now.


In order, 
the libusb_open_device_with_vid_pid function finds my card well,
the function libusb_claim_interface claim well my card,
the libusb_bulk_transfer function in usb_write returns me a -7 and therefore a timeout,
finally the libusb_bulk_transfer function in usb_read also returns me a -7 so again a timeout.



My C code bulk.cpp : 

      #include "bulk.h"
      #include <stdio.h>
      #include <stdlib.h>
      
      #include <signal.h>
      #include <libusb.h>
      
      
      static libusb_context* ctx = NULL;
      static libusb_device_handle* handle;
      
      static uint8_t receiveBuf[12];
      unsigned char transferBuf[12]="HELLO WORLD";
      
      
      
      /*
       * Read a packet
       */
      int usb_read(void)
      {
	      int nread, ret;
	      ret = libusb_bulk_transfer(handle, USB_ENDPOINT_IN, receiveBuf, sizeof(receiveBuf),&nread, USB_TIMEOUT);
	      if (ret) {
      
		      //ERROR
		      return -1;
	      }
	      else {
		      return 0;
	      }
      }
      
      
      /*
       * write a few bytes to the device
       *
       */
      int usb_write(void)
      {
	      int n, ret;
	      n = 0;
      
	      ret = libusb_bulk_transfer(handle, USB_ENDPOINT_OUT, transferBuf,sizeof(transferBuf), &n, USB_TIMEOUT);
      
	      //Error handling
	      switch (ret) {
	      case 0:
		      printf("send %d bytes to device\n", n);
		      return 0;
	      case LIBUSB_ERROR_TIMEOUT:
		      printf("ERROR in bulk write: %d Timeout\n", ret);
		      break;
	      case LIBUSB_ERROR_PIPE:
		      printf("ERROR in bulk write: %d Pipe\n", ret);
		      break;
	      case LIBUSB_ERROR_OVERFLOW:
		      printf("ERROR in bulk write: %d Overflow\n", ret);
		      break;
	      case LIBUSB_ERROR_NO_DEVICE:
		      printf("ERROR in bulk write: %d No Device\n", ret);
		      break;
	      default:
		      printf("ERROR in bulk write: %d\n", ret);
		      break;
      
	      }
	      return -1;
      
      }
      
      int test_bulk(void) 
      {
      
	      libusb_context* context = NULL;
	      libusb_device** list = NULL;
	      int rc = 0;
	      ssize_t count = 0;
      
	      libusb_init(&ctx);
      
	      //Open Device with VendorID and ProductID
	      handle = libusb_open_device_with_vid_pid(ctx,
		      USB_VENDOR_ID, USB_PRODUCT_ID);
	      if (!handle) {
		      return BULK_NOT_FOUND;
	      }
      
	      int ret = 1;
      
	      //Claim Interface 0 from the device
	      ret = libusb_claim_interface(handle, 0);
	      if (ret < 0) {
		      return BULK_INTERFACE_NOT_CLAIM;
	      } 
	      
	      while (1) 
	      {
		      usb_write();
		      usb_read();
	      }
	      libusb_close(handle);
	      libusb_exit(NULL);
      
      
	      return BULK_SUCCEED;
      }

My bulk.h:

        #define USB_VENDOR_ID	    0x0483      /* USB vendor ID used by the device
										         * 0x0483 is STMs ID
										         */
        #define USB_PRODUCT_ID	    0x209    /* USB product ID used by the device */
        #define USB_ENDPOINT_IN	    0x83   /* endpoint address */
        #define USB_ENDPOINT_OUT	0x3   /* endpoint address */
        #define USB_TIMEOUT	        3000        /* Connection timeout (in ms) */
        
        enum state{
	        BULK_NOT_FOUND,
	        BULK_INTERFACE_NOT_CLAIM,
	        BULK_SUCCEED
        };
        
        int usb_read(void);
        
        int test_bulk(void);
        
        static int usb_write(void);

**Reply by manuelbl** [reply]
The Windows code looks good for this limited example. The timeout in usb_write() can have multiple reasons:

- The device is not currently accepting data. It might have crashed or the buffers are full (possibly from a previous run). Unplug and replug it and ensure it hasn't crashed.
- The endpoint address is incorrect.

I assume the device has the WinUSB driver installed.

Once you get beyond these timeout issue, I propose two changes.

On the device:

```
void bulk_task(void) {
    int nread = tud_vendor_available();
    int nwrite = tud_vendor_write_available();
    int n = min(nread, nwrite);
    if (n == 0)
        return;

    uint8_t buf[64];
    n = min(n, sizeof(buf));
    tud_vendor_read(buf, n);
    tud_vendor_write(buf, n);
    tud_vendor_flush();
}
```

That way, you have proper flow control: no data lost, no overflows.

And for the host side:

Reading only 12 bytes works if you are sure that no more data can possibly arrive. If more data is available, an error will occur (something like *buffer overrun*). The libusb transfer in API and many other USB API do not behave like read() for a file. They merge USB packets but they don't like to split USB packets. The best approach is to specify a buffer that fits at least the maximum packet size (64 bytes for full speed, 512 bytes for high speed).
